src/gpgpu-sim/mem_fetch.h

```diff
@@ -105,6 +105,7 @@ class mem_fetch {
   unsigned get_timestamp() const { return m_timestamp; }
   unsigned get_return_timestamp() const { return m_timestamp2; }
   unsigned get_icnt_receive_time() const { return m_icnt_receive_time; }
+  unsigned long long get_streamID() const { return m_inst.get_streamid(); }
```

Comment:
m_inst could be NULL I think. Did you check every constructor of mem_fetch?

Reply:
A example is in gpu-cache.cc::1433

Reply:
You must mean some other line @JRPan ? https://github.com/accel-sim/gpgpu-sim_distribution/blob/dev/src/gpgpu-sim/gpu-cache.cc#L1443

Reply:
My bad. My branch line number is not the same.  Should have posted a link.
https://github.com/accel-sim/gpgpu-sim_distribution/blob/dev/src/gpgpu-sim/gpu-cache.cc#L1475

This should be the line I'm talking about. The second field of the mem_fetch constructor is NULL.

Reply:
Fair point ... I don't have a good solution for this.  I guess we'd have to pass streamID into this function?

Reply:
What do you think about storing stream_id at `mem_fetch()`, Instead of the `m_inst` inside `mem_fetch()`? 
That's how I did it. Is there any specific reason you save stream id in the m_inst instead of mem_fetch?

Reply:
The way we are having information flow through the system about stream ID, it's coming all the way from the Accel-Sim interface.  mem_fetches are created way later, and we needed a way to pass this information into the mem_fetch since multiple kernels might be running on the same SM.  Hence, we store the information in the instruction when it's created since we know what kernel/stream it came from then (and don't know later).

Reply:
Oh yes, you are correct. I also added the streamID to m_inst as you did. But I added streamID to the `mem_fetch()` constructor also. So I'm storing the streamID in both the mem_fetch and the m_inst. Then at the construction of `n_mf`, I can pass in the streamID of `mf`. 
The streamID is actually available there actually (https://github.com/accel-sim/gpgpu-sim_distribution/blob/dev/src/gpgpu-sim/gpu-cache.cc#L1475). The original `mf` (https://github.com/accel-sim/gpgpu-sim_distribution/blob/dev/src/gpgpu-sim/gpu-cache.cc#L1389) should have a valid m_inst, thus a valid streamID. 
I'm unsure about using the original m_inst (the one in mf) to construct the new mf (n_mf). Because technically, the `n_mf` is not from the original inst. Functionally it should be fine tho. What do you think?

Reply:
As long as the original mem_fetch object can be accessed, we can get stream ID from there.  But probably this will require we update this declaration to take stream ID in -- since otherwise it will be too late.

Reply:
Yes I agree. In my implementation, I changed the mem_fetch constructor and requires a streamID always. I can merge that so you don't have to reinvent the wheel. 
Or you can add streamID to the mem_fetch object, and at here https://github.com/accel-sim/gpgpu-sim_distribution/blob/dev/src/gpgpu-sim/mem_fetch.cc#L48 add a line like `m_streamID = m_inst->get_streamID()`. Then add another constructor that does not have a inst field, but have a streamID field to set the streamID directly, and leave `m_inst = NULL`. Then in gpu-cache.cc change every mem_fetch that has a NULL inst field to use the new mem_fetch constructor and pass the streamID from the original mf. My explanation could be bad. Does it make sense? I'm fine with either way.

Reply:
Why don't we just change 1475 to pass mf->inst instead of NULL?

Reply:
Because the m_inst generates the original mf, and then the cache generated n_mf to fetch the data on miss. There is no inst for the n_mf created. But as I said, functionally this should be fine. It's okay if you do that. You can make the call.

Reply:
> Yes I agree. In my implementation, I changed the mem_fetch constructor and requires a streamID always. I can merge that so you don't have to reinvent the wheel. Or you can add streamID to the mem_fetch object, and at here https://github.com/accel-sim/gpgpu-sim_distribution/blob/dev/src/gpgpu-sim/mem_fetch.cc#L48 add a line like `m_streamID = m_inst->get_streamID()`. Then add another constructor that does not have a inst field, but have a streamID field to set the streamID directly, and leave `m_inst = NULL`. Then in gpu-cache.cc change every mem_fetch that has a NULL inst field to use the new mem_fetch constructor and pass the streamID from the original mf. My explanation could be bad. Does it make sense? I'm fine with either way.

I tried the suggested separate constructor above. However, "new mem_fetch" was also called in l2cache.cc and shader.cc with inst as NULL, which caused compile problem: "error: call of overloaded ‘mem_fetch(mem_access_t&, NULL, int, int, int, int, const memory_config* const&, long long unsigned int&)’ is ambiguous". I do not think any "old" memfetch are available at those invocations. How were you treating them?

Reply:
Could I just make const warp_inst_t *inst an optional arg with a default value of NULL, and remove the NULL in the current calls?

Reply:
I added a field in `partition_mf_allocator::alloc` parameter to include stream id. Then pass the id when calling `new mem_fetch`. This can be complicated as the function is overloaded in several places. Below is how I did it. Just for your reference so you have a better understanding of all these. I can take care of it. Just find a way to get past it, so it's not blocking you, and add a comment to remind me. 
Of course, if you want you can do it. But I don't want to waste your time on this. You've done a lot and I really appreciate that.

Search for `shader_core_mem_fetch_allocator` in `shader.h`, and change `alloc` to add stream_id. And change `shader.cc` correspondingly. 
Then do the same for l2. Search for `partition_mf_allocator` in `l2cache.h`. Then update `l2cache.cc`
 Search for `mem_fetch_allocator` in `abstract_hardware_model.h` and make the changes. And also update the `abstract_hardware_model.cc`
At last change all calls to the `alloc` to pass a stream_id. It should always have one.

Reply:
Addressed in commit dc67564